File: Algorithms/Algo_Lab_Sheet_20/lab_Sheet_Two.py

import time
import matplotlib.pyplot as plt
import random

def matrix_chain_multiplication (array) :
    h = len (array)
    # Create a DP table to store results of subproblems
    dp = [ [ 0 for _ in range (h) ] for _ in range (h) ]
    
    # Solve subproblems for chains of length (from 2 to h-1)
    for l in range(2, h):  # l is the chain length
        for i in range(1, (h - l + 1)):
            j = (i + l - 1)
            dp[i][j] = float ('inf')  # Initialize with infinity
            for k in range (i, j) :
                # Calculate cost of splitting at k
                cost = (dp[i][k] + dp[k+1][j] + array[i-1] * array[k] * array[j])
                if cost < dp[i][j] :
                    dp[i][j] = cost
    
    return (dp[1][h-1])

# Measure time taken for different input sizes
def measure_time():
    input_sizes = [3, 4, 5, 6, 7, 8, 9, 10]
    times = []
    
    for size in input_sizes :
        # Generate random but valid matrix dimensions
        array = [random.randint (10, 100) for _ in range (size)]
        
        # Ensure the array has at least two elements and forms valid matrix dimensions
        if len (array) < 2 :
            print("Error :  Array size must be at least 2.")
            continue
        
        # Measure execution time
        start_time = time.time ()
        result = matrix_chain_multiplication (array)
        end_time = time.time ()
        elapsed_time = (end_time - start_time)
        
        print(f"Input Size :  {size}, Time Taken :  {elapsed_time:.6f} seconds, Result :  {result}")
        times.append (elapsed_time)
    
    # Plot graph
    plt.plot (input_sizes, times, marker='o')
    plt.xlabel ("Input Size")
    plt.ylabel ("Time Taken (seconds)")
    plt.title ("Matrix Chain Multiplication: Input Size vs Time Taken")
    plt.grid (True)
    plt.show ()

# Run the experiment
if __name__ == "__main__":
    measure_time()

File: Algorithms/Algo_Lab_Sheet_20/lab_Sheet_Two.java

import java.util.Random;

public class lab_Sheet_Two 
{
    // Function to calculate the minimum number of scalar multiplications
    public static int matrixChainMultiplication (int [] array) 
    {
        int h = array.length;
        // Create a DP table to store results of subproblems
        int [][] dp = new int [h][h];

        // Solve subproblems for chains of length (from 2 to h-1)
        for (int l = 2; l < h; l++) 
        {   // l is the chain length
            for (int i = 1; i < (h - l + 1); i++) 
            {
                int j = (i + l - 1);
                dp [i][j] = Integer.MAX_VALUE; // Initialize with infinity
                for (int k = i; k < j; k++) 
                {
                    // Calculate cost of splitting at k
                    int cost = (dp[i][k] + dp[k + 1][j] + array[i - 1] * array[k] * array[j]);
                    if (cost < dp[i][j]) 
                    {
                        dp[i][j] = cost;
                    }
                }
            }
        }
        return (dp[1][h - 1]); // Result for the full chain
    }

    // Function to measure time taken for different input sizes
    public static void measureTime () 
    {
        int [] inputSizes = {3, 4, 5, 6, 7, 8, 9, 10};
        Random rand = new Random ();

        for (int size : inputSizes) 
        {
            // Generate random matrix dimensions
            int [] array = new int [size];
            for (int i = 0; i < size; i++) 
            {
                array[i] = rand.nextInt(90) + 10; // Random values between 10 and 100
            }

            // Measure execution time
            long startTime = System.nanoTime ();
            int result = matrixChainMultiplication (array);
            long endTime = System.nanoTime ();
            double elapsedTime = (endTime - startTime) / 1_000_000_000.0; // Convert to seconds

            System.out.printf("Input Size :  %d, Time Taken :  %.6f seconds, Result :  %d%n",
                              size, elapsedTime, result);
        }
    }

    // Main function
    public static void main(String[] args) 
    {
        measureTime();
    }
}
